src/init/features/apphosting/index.ts

```diff
@@ -161,6 +162,83 @@ function defaultComputeServiceAccountEmail(projectId: string): string {
   return `${DEFAULT_COMPUTE_SERVICE_ACCOUNT_NAME}@${projectId}.iam.gserviceaccount.com`;
 }
 
+function defaultCloudBuildServiceAccount(projectNumber: string): string {
+  return `${projectNumber}@cloudbuild.gserviceaccount.com`;
+}
+
+function defaultComputeEngineServiceAccount(projectNumber: string): string {
+  return `${projectNumber}-compute@developer.gserviceaccount.com`;
+}
+
+function fetchServiceAccounts(projectNumber: string): {
+  buildServiceAccount: string;
+  runServiceAccount: string;
+} {
+  // TODO: For now we will always return the default CBSA and CESA. When the getBackend call supports returning
+  // the attached service account in a given backend/location then return that value instead.
+  // Sample Call: await apphosting.getBackend(projectId, location, backendId); & make this function async
+  return {
+    buildServiceAccount: defaultCloudBuildServiceAccount(projectNumber),
+    runServiceAccount: defaultComputeEngineServiceAccount(projectNumber),
+  };
+}
+
+/**
+ * Grants the corresponding service accounts the necessary access permissions to the provided secret.
+ */
+export async function grantSecretAccess(
+  secretName: string,
+  location: string,
+  backendId: string,
+  projectId: string,
+  projectNumber: string,
+): Promise<void> {
+  const isExist = await secretManager.secretExists(projectId, secretName);
+  if (!isExist) {
+    throw new FirebaseError(`Secret ${secretName} does not exist in project ${projectId}`);
+  }
+
+  let serviceAccounts = { buildServiceAccount: "", runServiceAccount: "" };
+  try {
+    serviceAccounts = fetchServiceAccounts(projectNumber);
+  } catch (err: any) {
+    throw new FirebaseError(
+      `Failed to get backend ${backendId} at location ${location}. Please check the parameters you have provided.`,
+      { original: err },
+    );
+  }
+
+  const secret: secretManager.Secret = {
+    projectId: projectId,
+    name: secretName,
+  };
+
+  const bindings: iam.Binding[] = [
+    {
+      role: "roles/secretmanager.secretAccessor",
+      members: [
+        `serviceAccount:${serviceAccounts.buildServiceAccount}`,
+        `serviceAccount:${serviceAccounts.runServiceAccount}`,
+      ],
+    },
+    {
+      role: "roles/secretmanager.viewer",
+      members: [`serviceAccount:${serviceAccounts.buildServiceAccount}`],
+    },
+  ];
+
+  try {
+    await secretManager.setIamPolicy(secret, bindings);
```

Comment:
have you tested this out in prod? I'm pretty sure you want to merge the new bindings into the existing binding before setting. Otherwise you overwrite existing grants

Reply:
Can confirm. Consider adding helper methods like those used in src/gcp/cloudfunctions.ts (setInvokerCreate and setInvokerUpdate, where the former blindly writes and the latter merges)

Reply:
Yeah good call, I tested it but it was with some inherited permissions that couldn't be overridden. Fixed